### PR TITLE
Work around PHP bug

### PR DESCRIPTION
**Pull request by nicolas-grekas**
See https://3v4l.org/bPFGl and https://bugs.php.net/78122

### CODE REVIEW

File: lib/Doctrine/Common/Lexer/AbstractLexer.php [hunk 1]
```diff
@@ -258,6 +258,11 @@ protected function scan($input)
         $flags = PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_OFFSET_CAPTURE;
         $matches = preg_split($regex, $input, -1, $flags);
 
+        if (false === $matches) {
+            // Work around https://bugs.php.net/78122
+            $matches = array(array($input, 0));
```
Comment (greg0ire) [inline lib/Doctrine/Common/Lexer/AbstractLexer.php]: Should we trigger a deprecation telling that input that triggers error described by `preg_last_error` will not be supported (as in, we throw an exception) in the next major version of the lexer? Or are we supposed to always allow this situation?

Comment (alcaeus) [reply lib/Doctrine/Common/Lexer/AbstractLexer.php]: I don't think that's necessary: there was talk about replacing `doctrine/lexer` with HOA parser in `doctrine/annotations` and `doctrine/dbal`. We can always add the deprecation in 1.1 if we decide to do a 2.0 release.

Comment (greg0ire) [reply lib/Doctrine/Common/Lexer/AbstractLexer.php]: Oh ok I didn't know this package was going to be abandoned. I still think a deprecation should be thrown as a way to tell the user they are probably misusing that method and that bad things may happen. That would mean a deprecation w/o any hint about a 2.0 version of the package:
```suggestion
            @trigger_error(sprintf(
                'input "%s" caused preg_split() to fail with error code %d (description at https://www.php.net/manual/en/pcre.constants.php). "%s" may not behave as expected',
                $input,
                preg_last_error(),
                __METHOD__
            ), E_USER_DEPRECATED);
            $matches = array(array($input, 0));
```

**Comment by greg0ire** [comment]
> Work around PHP bug

Technically not a bug as the link shows, it's more of a bugfix that is not BC, if I understand correctly.

Also, can someone please explain to me why this is needed? I'm not familiar with the lexer, I don't even know what it is lexing… DQL, probably? If this is it, in what situation would it be normal to have that kind of weird single-byte characters in a DQL string? And is the current behavior of the lexer correct in that case.

**Comment by SenseException** [comment]
Also: `[]` instead of `array()` syntax.

**Comment by nicolas-grekas** [comment]
> Also: [] instead of array() syntax.

Please check the CS of the file. I'm just keeping it consistent.

@greg0ire this behavior change breaks the Symfony Validator test suite, in relation with the egulas strict email parser which uses this lexer.

Your concern are for another PR to me.

**Comment by greg0ire** [comment]
> Your concerns are for another PR to me.

Indeed, let's keep this about restoring the previous behavior

**Comment by greg0ire** [comment]
Here are the errors:

```
1) Symfony\Component\Validator\Tests\Constraints\EmailValidatorTest::testStrictWithInvalidEmails with data set #26 ('�@iana.org')
Invalid argument supplied for foreach()

/home/greg/dev/symfony/vendor/doctrine/lexer/lib/Doctrine/Common/Lexer/AbstractLexer.php:261
/home/greg/dev/symfony/vendor/doctrine/lexer/lib/Doctrine/Common/Lexer/AbstractLexer.php:96
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/EmailParser.php:41
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/Validation/RFCValidation.php:30
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/Validation/NoRFCWarningsValidation.php:21
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/EmailValidator.php:37
/home/greg/dev/symfony/src/Symfony/Component/Validator/Constraints/EmailValidator.php:79
/home/greg/dev/symfony/src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:244

2) Symfony\Component\Validator\Tests\Constraints\EmailValidatorTest::testStrictWithInvalidEmails with data set #27 ('test@�.org')
Invalid argument supplied for foreach()

/home/greg/dev/symfony/vendor/doctrine/lexer/lib/Doctrine/Common/Lexer/AbstractLexer.php:261
/home/greg/dev/symfony/vendor/doctrine/lexer/lib/Doctrine/Common/Lexer/AbstractLexer.php:96
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/EmailParser.php:41
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/Validation/RFCValidation.php:30
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/Validation/NoRFCWarningsValidation.php:21
/home/greg/dev/symfony/vendor/egulias/email-validator/EmailValidator/EmailValidator.php:37
/home/greg/dev/symfony/src/Symfony/Component/Validator/Constraints/EmailValidator.php:79
/home/greg/dev/symfony/src/Symfony/Component/Validator/Tests/Constraints/EmailValidatorTest.php:244
```

**Comment by alcaeus** [comment]
Released as part of 1.0.2. Thanks @nicolas-grekas!

**Pull request merged**